### DISCUSSION TITLE
my normalization implementation

### DISCUSSION THREAD

**Discussion by gpawru** [unicode-org/icu4x#4647] [Show and tell]
Hello everyone!

First of all, thank you very much for what you do, it's inspiring! :love_you_gesture: 

Some time ago, I decided to write a few articles about Unicode, and logically, I wrote several code examples for these articles. More specifically, I implemented my own normalization (in progress - collations; articles not yet published - being reviewed by the editor, but that's all off-topic).

By chance, I noticed this section and decided to share my results (maybe it will be useful to someone?). I would like to ask for advice: perhaps there is something worth correcting?  Should I make a crate out of these drafts?
Comments in the code are in russian (since the main purpose was to supplement the articles), but Google translate will handle it :)

I decided to sacrifice a bit of data size for performance. Here are the sizes of the compressed data:

- NFD: 69 Kb
- NFKD: 96 Kb
- NFC: 82 Kb
- NFKC: 113 Kb

Tests used: UCD tests + results comparsion with ICU4X; all combinations tests on early stages (not included in the repositories, SLOW, but helped to discover https://github.com/unicode-org/icu4x/discussions/4527 )

For benchmarks, I took texts in various languages (most common) (public domain texts, for some languages - simply Google translated), trimmed to 100 Kb, and made two versions: regular and predecomposed.

And finally the benchmarks:

### Decomposing, µs :

| language   | ICU4X NFD | my NFD | ICU4X NFKD | my NFKD | ICU4X NFD (dec[^1]) | my NFKD (dec[^1]) |
| ---------- | --------- | ------ | ---------- | ------- | --------------- | ------------- |
| arabic | 1244 | 508 | 1631 | 515 | 1230 | 508 |
| chinese | 883 | 201 | 1627 | 355 | 890 | 221 |
| czech | 886 | 603 | 992 | 603 | 875 | 558 |
| dutch | 178 | 111 | 191 | 108 | 178 | 108 |
| english | 193 | 110 | 214 | 109 | 188 | 104 |
| french | 360 | 228 | 385 | 233 | 337 | 208 |
| german | 249 | 163 | 283 | 166 | 235 | 149 |
| greek | 1473 | 713 | 1879 | 708 | 1516 | 641 |
| hebrew | 1073 | 407 | 1419 | 406 | 1072 | 402 |
| hindi | 867 | 341 | 1164 | 341 | 861 | 341 |
| italian | 257 | 158 | 287 | 158 | 245 | 145 |
| japanese | 1072 | 321 | 1665 | 337 | 1047 | 284 |
| korean | 2662 | 1167 | 3277 | 1261 | 2131 | 595 |
| persian | 1127 | 443 | 1511 | 434 | 1132 | 421 |
| polish | 653 | 442 | 755 | 442 | 635 | 421 |
| portuguese | 287 | 186 | 304 | 182 | 275 | 170 |
| russian | 1115 | 420 | 1508 | 425 | 1111 | 406 |
| serbian | 1045 | 411 | 1472 | 411 | 1063 | 412 |
| spanish | 329 | 206 | 368 | 205 | 304 | 189 |
| thai | 904 | 345 | 1307 | 386 | 905 | 342 |
| turkish | 738 | 500 | 855 | 499 | 708 | 466 |
| ukrainian | 1115 | 421 | 1574 | 446 | 1099 | 412 |
| vietnamese | 1923 | 1182 | 2252 | 1304 | 1673 | 1111 |

### Composing, µs :

| language   | ICU4X NFC | my NFC | ICU4X NFKC | my NFKC | ICU4X NFC (dec[^1]) | my NFKC (dec[^1]) |
| ---------- | --------- | ------ | ---------- | ------- | --------------- | ------------- |
| arabic | 890 | 411 | 1359 | 443 | 1130 | 574 |
| chinese | 787 | 203 | 1738 | 471 | 772 | 205 |
| czech | 104 | 71 | 592 | 334 | 2047 | 1045 |
| dutch | 104 | 72 | 164 | 113 | 111 | 73 |
| english | 156 | 117 | 180 | 109 | 158 | 120 |
| french | 104 | 84 | 238 | 152 | 603 | 317 |
| german | 102 | 72 | 198 | 122 | 286 | 172 |
| greek | 970 | 383 | 1431 | 425 | 1864 | 879 |
| hebrew | 855 | 425 | 1302 | 457 | 848 | 420 |
| hindi | 848 | 395 | 1129 | 432 | 842 | 399 |
| italian | 136 | 94 | 204 | 128 | 305 | 171 |
| japanese | 859 | 194 | 1498 | 261 | 1156 | 345 |
| korean | 874 | 312 | 1425 | 377 | 4468 | 2605 |
| persian | 823 | 404 | 1270 | 433 | 893 | 446 |
| polish | 105 | 74 | 495 | 269 | 1012 | 583 |
| portuguese | 105 | 73 | 195 | 129 | 449 | 244 |
| russian | 794 | 403 | 1272 | 447 | 839 | 430 |
| serbian | 788 | 428 | 1243 | 453 | 778 | 419 |
| spanish | 119 | 82 | 230 | 144 | 506 | 250 |
| thai | 1025 | 462 | 1431 | 531 | 1017 | 457 |
| turkish | 111 | 77 | 585 | 307 | 1041 | 595 |
| ukrainian | 790 | 384 | 1271 | 430 | 912 | 450 |
| vietnamese | 564 | 297 | 1202 | 575 | 3761 | 2061 |

[^1]: dec - predecomposed texts used

### Repositories:

- decomposing: https://github.com/gpawru/02_habr_decomposing_normalization
- composing: https://github.com/gpawru/02_habr_composing_normalization
- UCD parse: https://github.com/gpawru/unicode_data
- baking data: https://github.com/gpawru/unicode_bakery

**Comment by sffc** [comment]
Thanks for the note!

Do you have a sense of where the performance difference might come from? Is it just that your data size is larger (100 kB is quite large)? I would be curious to see if you can improve on ICU4X performance with smaller data files. Alternatively, it would be compelling to see numbers if you can share data between the four normalization routines instead of shipping big data for each one.

**Reply by gpawru** [reply]
@sffc Hello Shane!

Now:

* NFD - 29 Kb,
* NKFD - 56 Kb.
* Compositions - common dataset - 14 Kb.
* NFC = NFD + Compositions + 104 bytes patch
* NFKC = NFKD + Compositions + 1288 bytes patch

Performance in benchmarks, of course, has decreased, but not critically. Benchmark result tables are [here](https://github.com/gpawru/02_habr_decomposing_normalization/blob/master/README.md) and [here](https://github.com/gpawru/02_habr_composing_normalization/blob/master/README.md)
